experimental/sql/dialects/rel_ssa.py

```diff
@@ -0,0 +1,320 @@
+# Licensed under the Apache License v2.0 with LLVM Exceptions.
+# See https://llvm.org/LICENSE.txt for license information.
+# SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+from dataclasses import dataclass
+from typing import Any, Union, List
+from xdsl.ir import Block, Region, Operation, SSAValue, ParametrizedAttribute, Data, MLContext, Attribute
+from xdsl.dialects.builtin import StringAttr, ArrayAttr, ArrayOfConstraint, IntegerAttr
+from xdsl.irdl import AttributeDef, OperandDef, ResultDef, RegionDef, SingleBlockRegionDef, irdl_attr_definition, irdl_op_definition, ParameterDef, AnyAttr, VarOperandDef, builder
+
+# This file contains the relational SSA dialect, a dialect that expresses
+# relational queries using SSA. The dialect consists of two types of operations:
+# Expressions and Operators. Operators work on Bags whereas expressions work on
+# single values and are used to specify the exact behavior of operators. The
+# Column and Yield Expressions bridge the gap to convert from one to the other.
+
+#===------------------------------------------------------------------------===#
+# Data types
+#===------------------------------------------------------------------------===#
+
+
+@irdl_attr_definition
+class DataType(ParametrizedAttribute):
+  """
+  Models a datatype in a relational SSA query.
+  """
+  name = "rel_ssa.datatype"
+
+
+@irdl_attr_definition
+class Int32(DataType):
+  """
+  Models a int32 type in a relational SSA query.
+
+  Example:
+
+  ```
+  !rel_ssa.int32
+  ```
+  """
+  name = "rel_ssa.int32"
+
+
+@irdl_attr_definition
+class String(DataType):
+  """
+  Models a string type in a relational SSA query, that can be either nullable or
+  not.
+
+  Example:
+
+  ```
+  !rel_ssa.string<0: !i1>
+  ```
+  """
+  name = "rel_ssa.string"
+
+  # TODO: redefine nullable as a property of all fields
+  nullable = ParameterDef(IntegerAttr)
```

Comment:
I am not completely clear what this is .. in core MLIR parlance, is this an IntegerAttr "type attribute" that describes whether the string is nullable or not (i.e. optional) ?

If so, it seems quite unfortunate that xdsl renames MLIR concepts and we should make it clear and loud in the comments at least that this is an MLIR Type Attribute.

Also, if my interpretation is correct, should this be a BoolAttr ?

Reply:
AFAIU, this is exactly the [MLIR IntegerAttr](https://mlir.llvm.org/docs/Dialects/Builtin/#integerattr). It means that this type has a parameter that defines whether it is nullable or not. And this parameter stores an integer.

I agree that `BoolAttr` would be nicer, but this has not yet been introduced to our builtin dialect since its basically sugar for i1 anyhow. In a final version, I agree that this should be a `BoolAttr`.